File: ChessMateGUI/src/main/java/com/flynn/chessmate/communication/Message.java
package com.flynn.chessmate.communication;

/**
 * Class to represent a message send over the wire. Contains a message type
 * and message data.
 *
 * @author Timothy Flynn (timothy.flynn@outlook.com)
 * @version March 3, 2013
 */
public class Message
{
    static private char CHESSMATE_EOM = 0x04;

    private MessageType m_type;
    private String m_data;

    // Enumeration of message types
    public enum MessageType {
        INVALID_TYPE(-1),
        START_GAME(0),
        INVALID_MOVE(1),
        MAKE_MOVE(2),
        GET_MOVE(3),
        DISCONNECT(4),
        NUM_TYPES(5);

        private final int m_val;

        MessageType(int val)
        {
            m_val = val;
        }

        public int getValue()
        {
            return m_val;
        }

        public String toString()
        {
            return String.valueOf(m_val);
        }
    }

    public static boolean isEoM(char character)
    {
        return (character == CHESSMATE_EOM);
    }

    /**
     * Default constructor to create an invalid message.
     */
    public Message()
    {
        m_type = MessageType.INVALID_TYPE;
        m_data = "";
    }

    /**
     * Constructor to determine type and data from a raw string.
     */
    public Message(String raw)
    {
        if (raw != null && raw.length() > 0)
        {
            m_type = MessageType.values()[(Integer.parseInt(raw.substring(0, 1))) + 1];

            m_data = raw.substring(2);
        }
        else
        {
            m_type = MessageType.INVALID_TYPE;
            m_data = "";
        }
    }

    /**
     * Constructor to stored a known type.
     */
    public Message(MessageType type)
    {
        m_type = type;
        m_data = "";
    }

    /**
     * Constructor to stored a known type and data.
     */
    public Message(MessageType type, String data)
    {
        m_type = type;
        m_data = data;
    }

    /**
     * @return The message's enumerated type.
     */
    public MessageType getMessageType()
    {
        return m_type;
    }

    /**
     * @return The message's data.
     */
    public String getData()
    {
        return m_data;
    }

    /**
     * Return this message as a string.
     */
    public String serialize()
    {
        return (m_type.toString() + " " + m_data + CHESSMATE_EOM);
    }
}

File: ChessMateEngine/game/message.cpp
#include "message.h"

#include <fly/types/string/string.hpp>
#include <fly/types/string/string_lexer.hpp>

#include <optional>

namespace chessmate {

namespace {

    char s_chessmate_eom = 0x04;

} // namespace

//==================================================================================================
Message::Message() : m_type(Message::INVALID_TYPE)
{
}

//==================================================================================================
Message::Message(const std::string &raw) : m_type(Message::INVALID_TYPE)
{
    using StringLexer = fly::BasicStringLexer<std::string>;

    StringLexer lexer(raw);

    if (auto type = lexer.consume_number(); type)
    {
        m_type = static_cast<Message::MessageType>(type.value());

        if (!lexer.consume_if(' '))
        {
            return;
        }
        else if (lexer.position() >= raw.size())
        {
            return;
        }

        m_data = raw.substr(lexer.position(), lexer.view().size() - lexer.position() - 1);
    }
}

//==================================================================================================
Message::Message(Message::MessageType type, const std::string &data) : m_type(type), m_data(data)
{
}

//==================================================================================================
bool Message::IsValid() const
{
    bool isValid = false;

    switch (m_type)
    {
        // START_GAME data of the form "<engine color> <difficulty>"
        case Message::START_GAME:
            isValid = (m_data.length() > 2);
            break;

        // MAKE_MOVE, INVALID_MOVE data of the form "<PGN string> [stalemate]"
        case Message::INVALID_MOVE:
        case Message::MAKE_MOVE:
            isValid = (m_data.length() > 0);
            break;

        // GET_MOVE, DISCONNECT have no data
        case Message::GET_MOVE:
        case Message::DISCONNECT:
            isValid = (m_data.length() == 0);
            break;

        default:
            break;
    }

    return isValid;
}

//==================================================================================================
Message::MessageType Message::GetMessageType() const
{
    return m_type;
}

//==================================================================================================
std::string Message::GetData() const
{
    return m_data;
}

//==================================================================================================
std::string Message::Serialize() const
{
    return fly::String::format(
        "{}{}{}",
        m_type,
        (m_data.empty() ? "" : " " + m_data),
        s_chessmate_eom);
}

} // namespace chessmate
